### ISSUE TITLE
As the number of training rounds increases, the values of cost and reward change irregularly.

### ISSUE THREAD

**Issue by 623637719** [liuzuxin/OSRL#22]
In my env,as the number of training rounds increases, the values of cost and reward change irregularly,The training results are shown below,it doesn't look normal.
![image](https://github.com/liuzuxin/OSRL/assets/111236370/a9c2d358-953a-4d45-8f21-794ad257d9f8)
![image](https://github.com/liuzuxin/OSRL/assets/111236370/9bbc3390-9fd9-4635-bf28-b4ccd94005c4)
![image](https://github.com/liuzuxin/OSRL/assets/111236370/59f059bd-94c7-4f5b-a301-043d4ba0bcac)

However,When I choose other env(like OfflinePointCircle2Gymnasium-v0), the cost gradually declines as the number of training rounds increases, the ret rises as the number of training rounds increases,it look normal.

Here's the parameters in my cdt_config.py
```
@dataclass
class CDTTrainConfig:
    # wandb params
    project: str = "OSRL-baselines"
    group: str = None
    name: Optional[str] = None
    prefix: Optional[str] = "CDT"
    suffix: Optional[str] = ""
    logdir: Optional[str] = "logs"
    verbose: bool = True
    # dataset params
    outliers_percent: float = None
    noise_scale: float = None
    inpaint_ranges: Tuple[Tuple[float, float], ...] = None
    epsilon: float = None
    density: float = 1.0
    # model params
    embedding_dim: int = 128
    num_layers: int = 3
    num_heads: int = 8
    action_head_layers: int = 1
    seq_len: int = 10
    episode_len: int = 300
    attention_dropout: float = 0.1
    residual_dropout: float = 0.1
    embedding_dropout: float = 0.1
    time_emb: bool = True
    # training params
    #task: str = "OfflinePointCircle2Gymnasium-v0"
    task: str = "Autobidding-v0"
    dataset: str = None
    learning_rate: float = 1e-4
    betas: Tuple[float, float] = (0.9, 0.999)
    weight_decay: float = 1e-4
    clip_grad: Optional[float] = 0.25
    batch_size: int = 8
    update_steps: int = 5000
    lr_warmup_steps: int = 200
    reward_scale: float = 0.1
    cost_scale: float = 1
    num_workers: int = 0
    # evaluation params
    target_returns: Tuple[Tuple[float, ...],
                          ...] = ((450.0, 10), (500.0, 20), (550.0, 50))  # reward, cost
    #The cost limit corresponds to the cost threshold for your problem, it should be the same as your target cost return for CDT.
    cost_limit: int = 100
    eval_episodes: int = 10
    eval_every: int = 100
    # general params
    seed: int = 0
    device: str = "cuda:0"
    threads: int = 6
    # augmentation param
    deg: int = 4
    pf_sample: bool = False
    beta: float = 1.0
    augment_percent: float = 0.2
    # maximum absolute value of reward for the augmented trajs
    max_reward: float = 1.0
    # minimum reward above the PF curve
    min_reward: float = 0.2
    # the max drecrease of ret between the associated traj
    # w.r.t the nearest pf traj
    max_rew_decrease: float = 0.5
    # model mode params
    use_rew: bool = True
    use_cost: bool = True
    cost_transform: bool = True
    cost_prefix: bool = False
    add_cost_feat: bool = False
    mul_cost_feat: bool = False
    cat_cost_feat: bool = False
    loss_cost_weight: float = 0.02
    loss_state_weight: float = 0
    cost_reverse: bool = False
    # pf only mode param
    pf_only: bool = False
    rmin: float = 300
    cost_bins: int = 60
    npb: int = 5
    cost_sample: bool = True
    linear: bool = True  # linear or inverse
    start_sampling: bool = False
    prob: float = 0.2
    stochastic: bool = True
    init_temperature: float = 0.1
    no_entropy: bool = False
    # random augmentation
    random_aug: float = 0
    aug_rmin: float = 400
    aug_rmax: float = 500
    aug_cmin: float = -2
    aug_cmax: float = 25
    cgap: float = 5
    rstd: float = 1
    cstd: float = 0.2

@dataclass
class CDTCarCircleConfig(CDTTrainConfig):
    pass

@dataclass
class AutobiddingConfig(CDTTrainConfig):
    # model params
    seq_len: int = 10
    episode_len: int = 1000
    # training params
    task: str = "Autobidding-v0"
    target_returns: Tuple[Tuple[float, ...],
                          ...] = ((15.0, 20), (15.0, 40), (15.0, 80))# reward, cost
    # augmentation param
    deg: int = 1
    max_reward: float = 2
    min_reward: float = 1
    max_rew_decrease: float = 0.3
    device: str = "cuda:0"
```
The format of my data is shown below:
![image](https://github.com/liuzuxin/OSRL/assets/111236370/9f59d5cb-9be6-4d3f-b88d-c081576f5679)

May you help me?What do you think is the problem?

**Comment by liuzuxin** [comment]
I guess it is because the target reward and cost returns are not set properly. You should set

```
# evaluation params
target_returns: Tuple[Tuple[float, ...], ...] = ((450.0, 10), (500.0, 20), (550.0, 50))  # reward, cost
```
to be your datasets' corresponding reward return and cost return ranges.

**Comment by Ja4822** [comment]
Hi @623637719, I would suggest that you can (1) plot your dataset on the cost-reward return space as shown in Figures 10–14 in the [paper](https://arxiv.org/pdf/2306.09303.pdf) to determine the augmentation parameters such as `deg`, `max_reward`, `min_reward`, `max_rew_decrease` etc; (2) set `augment_percent` and `loss_cost_weight` to 0 and a smaller learning rate like 1e-4 to see if CDT works without data augmentation; (3) try larger `update_steps` and `eval_every`, I am not sure if CDT can converge in 5000 steps in your environment.

**Comment by 623637719** [comment]
> Hi @623637719, I would suggest that you can (1) plot your dataset on the cost-reward return space as shown in Figures 10–14 in the [paper](https://arxiv.org/pdf/2306.09303.pdf) to determine the augmentation parameters such as `deg`, `max_reward`, `min_reward`, `max_rew_decrease` etc; (2) set `augment_percent` and `loss_cost_weight` to 0 and a smaller learning rate like 1e-4 to see if CDT works without data augmentation; (3) try larger `update_steps` and `eval_every`, I am not sure if CDT can converge in 5000 steps in your environment.

Thank you for your answer！I have successfully drawn this image. After I adjust the augmentation parameters, how can I know the augmentation parameters I adjust are appropriate according to this image

**Comment by 623637719** [comment]
> We can find an example [here](https://arxiv.org/pdf/2302.07351.pdf) (Figure 3 and left plot in Figure 9). The augmentation parameters depend on the data. Generally, the fitted pareto frontier, let's say f, (by changing `deg`) shoud be monotonic. In our dataset, we set f(cmax)≤ `max_reward`, `min_reward` ≤f(cmin), where cmin,cmax are the maximum and minimum cost return in the dataset, so the augmented data are associated to diverse reward-cost pair in the original data. Besides, can you try (2) and (3) as I mentioned?

I'm sorry for my slow reply due to my recent internship。
I have set the "deg" parameter to 1, and looking at the figure below, it should be a good number to set.But that vertical dotted line seems a little too far to the right
![image](https://github.com/liuzuxin/OSRL/assets/111236370/3124708e-c7d5-4363-9766-bda69b4e67d3)

I have tried (2)and(3),but as measured by the training results, nothing has changed.

Then I checked the trainer.evaluate() part of the code and realized that I might have written the env.step() function wrong.
 I modified my env.step() function and noticed that the training results had changed. cost decreased with training, but ret did not increase significantly with training.
 
![image](https://github.com/liuzuxin/OSRL/assets/111236370/526f5805-5de2-4f76-bec8-a8c908835bfa)
![image](https://github.com/liuzuxin/OSRL/assets/111236370/7301906e-78e4-41be-84c5-517aa97e57d2)
![image](https://github.com/liuzuxin/OSRL/assets/111236370/7e581ce4-6190-47c3-8075-af6b957778db)
